src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -280,38 +307,111 @@ trait SlickExpressionParser extends JavaTokenParsers {
-  def evalFinalAttribute(workspaceContext: SlickWorkspaceContext, rootType:String, rootName:String, expression:String): Try[ReadAction[Iterable[AttributeValue]]] = {
-    parser.parseAttributeExpr(expression).map( runPipe(SlickExpressionContext(workspaceContext, rootType, rootName), _).map(_.collect {
-      case AttributeNull => Seq.empty
-      case AttributeEmptyList => Seq.empty
-      case av: AttributeValue => Seq(av)
-      case avl: AttributeValueList => avl.list
-      case badType => throw new RawlsException(s"unsupported type resulting from attribute expression: $badType: ${badType.getClass}")
-    }.flatten))
+object SlickExpressionEvaluator {
+  def withNewExpressionEvaluator[R](parser: DataAccess, rootEntities: Seq[EntityRecord])
+                                   (op: SlickExpressionEvaluator => ReadWriteAction[R])
+                                   (implicit executionContext: ExecutionContext): ReadWriteAction[R] = {
+    val evaluator = new SlickExpressionEvaluator(parser, rootEntities)
+
+    evaluator.populateExprEvalTempTable()
+      .andThen({ op(evaluator) })
+      .cleanUp({ _ => evaluator.dropExprEvalTempTable() })
   }
 
-  def evalFinalEntity(workspaceContext: SlickWorkspaceContext, rootType:String, rootName:String, expression:String): Try[ReadAction[Iterable[Entity]]] = {
-    parser.parseEntityExpr(expression)
-      .map( runPipe(SlickExpressionContext(workspaceContext, rootType, rootName), _).map(_.collect { case e: Entity => e } ))
+  def withNewExpressionEvaluator[R](parser: DataAccess, workspaceContext: SlickWorkspaceContext, rootType: String, rootName: String)
+                                   (op: SlickExpressionEvaluator => ReadWriteAction[R])
+                                   (implicit executionContext: ExecutionContext): ReadWriteAction[R] = {
+    import parser.driver.api._
+
+    //Find the root entity for the expression
+    val dbRootEntityRec = parser.entityQuery.findEntityByName(workspaceContext.workspaceId, rootType, rootName).result
+
+    //Sanity check that we've only got one, and then pass upwards
+    dbRootEntityRec flatMap { rootEntityRec =>
+      if(rootEntityRec.size != 1) {
+        DBIO.failed(new RawlsException(s"Found != 1 root entity when searching for ${rootType}/$rootName"))
+      } else {
+        withNewExpressionEvaluator(parser, rootEntityRec)(op)
+      }
+    }
+  }
+}
+
+class SlickExpressionEvaluator(val parser: DataAccess, val rootEntities: Seq[EntityRecord])(implicit executionContext: ExecutionContext) {
+  import parser.driver.api._
+
+  def populateExprEvalTempTable() = {
+    parser.exprEvalQuery ++= rootEntities.map( e => parser.ExprEvalRecord(e.name) )
+  }
+
+  def dropExprEvalTempTable() = {
+    parser.exprEvalQuery.delete
+  }
+
+  def evalFinalAttribute(workspaceContext: SlickWorkspaceContext, expression: String): ReadWriteAction[Map[String, Try[Iterable[AttributeValue]]]] = {
```

Comment:
The `DBIOAction` type is now `ReadWriteAction` because of the populate and drop of `EXPREVAL_TEMP`. Annoyingly, this propagates most of the way up the callstack.